realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -204,6 +204,14 @@ public RealmObjectSchema removeField(String fieldName) {
         if (table.getPrimaryKey() == columnIndex) {
             table.setPrimaryKey(null);
         }
+        // When PK does not exist (-2) this cannot happen. Meanwhile a removal target has a smaller
```

Comment:
It is true that removing a field does not affect PK stored in PK table. The removal however could generates index re-ordering of PK column if a column prior to PK gets removed. For example

```java
table.addField("a"); // index of 0
table.addField("b"); // index of 1
table.addField("c"); // index of 2
table.setPrimaryKey("b");
```
Now let's remove field `a`. 
```java
table.removeField("a");
```
Then `b`'s index is 0, and `c`'s index is 1 while `cachedPrimaryKeyColumnIndex` still has the value of 1 pointing to `c`. 

Then let's remove field `c`. 
```java
table.removeField("c");
```
What happens is `if (table.getPrimaryKey() == columnIndex)` becomes true and PK gets removed from PK table. `cachedPrimaryKeyColumnIndex` gets invalidated as well. 

**I was only looking for removing non-PK field `c` though**.

Whenever we are to remove a column with smaller index than that of PK field, we should recalculate `cachedPrimaryKeyColumnIndex` to properly point the correct PK field.

- - - 
So this what I did. 

When execution reaches at the point of `if (columnIndex < table.getPrimaryKey())`, there should be only two cases; 
1. PK is removed or does not exists so `table.getPrimaryKey()` will return -2. `columnIndex` should have proper value (i.e. 0 or bigger) that the condition should be false.
2. PK exists and has a bigger index value than `columnIndex` and this is exactly what we are looking for. So we go into the path recalculate `cachedPrimaryKeyColumnIndex` again and return.

- - - 
After writing down the detail, I can see I should have made the code more readable. I'll revise them.